packages/contract-utils/access-control/src/storage.rs

```diff
@@ -0,0 +1,534 @@
+///   privileges.
+///
+/// # Events
+///
+/// * topics - `["role_granted", role: Symbol, account: Address]`
+/// * data - `[caller: Address]`
+///
+/// # Notes
+///
+/// * Authorization for `caller` is required.
+pub fn grant_role(e: &Env, caller: &Address, account: &Address, role: &Symbol) {
+    caller.require_auth();
+    ensure_if_admin_or_admin_role(e, caller, role);
+
+    // Return early if account already has the role
+    if has_role(e, account, role).is_some() {
+        return;
+    }
+
+    let index = add_to_role_enumeration(e, account, role);
+
+    let key = AccessControlStorageKey::HasRole(account.clone(), role.clone());
+    e.storage().persistent().set(&key, &index);
+    e.storage().persistent().extend_ttl(&key, ROLE_TTL_THRESHOLD, ROLE_EXTEND_AMOUNT);
+
+    emit_role_granted(e, role, account, caller);
+}
+
+/// Revokes a role from an account.
+///
+/// # Arguments
+///
+/// * `e` - Access to Soroban environment.
+/// * `caller` - The address of the caller, must be the admin or has the
+///   `AdminRole` privileges for this role.
+/// * `account` - The account to revoke the role from.
+/// * `role` - The role to revoke.
+///
+/// # Errors
+///
+/// * `AccessControlError::Unauthorized` - If the `caller` does not have enough
+///   privileges.
+/// * `AccessControlError::AccountNotFound` - If the `account` doesn't have the
+///   role.
+/// * refer to [`remove_from_role_enumeration()`] errors.
+///
+/// # Events
+///
+/// * topics - `["role_revoked", role: Symbol, account: Address]`
+/// * data - `[caller: Address]`
+///
+/// # Notes
+///
+/// * Authorization for `caller` is required.
+pub fn revoke_role(e: &Env, caller: &Address, account: &Address, role: &Symbol) {
+    caller.require_auth();
+    ensure_if_admin_or_admin_role(e, caller, role);
+
+    // Check if account has the role
+    if has_role(e, account, role).is_none() {
+        panic_with_error!(e, AccessControlError::AccountNotFound);
+    }
+
+    remove_from_role_enumeration(e, account, role);
+
+    let key = AccessControlStorageKey::HasRole(account.clone(), role.clone());
+    e.storage().persistent().remove(&key);
+
+    emit_role_revoked(e, role, account, caller);
+}
+
+/// Allows an account to renounce a role assigned to itself.
+/// Users can only renounce roles for their own account.
+///
+/// # Arguments
+///
+/// * `e` - Access to Soroban environment.
+/// * `caller` - The address of the caller, must be the account that has the
+///   role.
+/// * `role` - The role to renounce.
+///
+/// # Errors
+///
+/// * `AccessControlError::AccountNotFound` - If the `caller` doesn't have the
+///   role.
+/// * refer to [`remove_from_role_enumeration()`] errors.
+///
+/// # Events
+///
+/// * topics - `["role_revoked", role: Symbol, account: Address]`
+/// * data - `[caller: Address]`
+///
+/// # Notes
+///
+/// * Authorization for `caller` is required.
+pub fn renounce_role(e: &Env, caller: &Address, role: &Symbol) {
```

Comment:
Is there anything that prevents there being no one with a role? This might not be something that should be prevented at at this layer, but I'm wondering if it is possible for a contract to get into a state where it no longer works. I'm specifically thinking about this because the `renounce_role` allows an individual account to modify the state of the overall contract. A developer using the role functionality, may set a contract's access control up, but not realise that any individual account can disrupt the overall state. I'm not thinking of preventing abuse so much as accidental misuse.

Reply:
Can the roles be used in isolation without an admin? That's what I thinking someone might do. Thinking of financial institutions that may not have a single superuser that can control all the funds in isolation.

Reply:
Roles can definitely be used in isolation without an admin. But the creation of the roles should be done with an admin. And admin can still revoke/grant roles afterwards. 
Do you mean, after the setup, admin should be able to renounce themselves?